src/embind/emval.js

```diff
@@ -12,49 +12,43 @@
 /*jslint sub:true*/ /* The symbols 'fromWireType' and 'toWireType' must be accessed via array notation to be closure-safe since craftInvokerFunction crafts functions as strings that can't be closured. */
 
 // -- jshint doesn't understand library syntax, so we need to mark the symbols exposed here
-/*global getStringOrSymbol, emval_handle_array, Emval, __emval_unregister, count_emval_handles, emval_symbols, emval_free_list, get_first_emval, __emval_decref, emval_newers*/
+/*global getStringOrSymbol, emval_handles, Emval, __emval_unregister, count_emval_handles, emval_symbols, __emval_decref, emval_newers*/
 /*global craftEmvalAllocator, emval_addMethodCaller, emval_methodCallers, LibraryManager, mergeInto, emval_allocateDestructors, global, emval_lookupTypes, makeLegalFunctionName*/
 /*global emval_get_global*/
 
 var LibraryEmVal = {
-  $emval_handle_array: [
-    {},
-    {value: undefined},
-    {value: null},
-    {value: true},
-    {value: false}
-  ], // reserve zero and special values
-  $emval_free_list: [],
+  $emval_handles__deps: ['$HandleAllocator'],
+  $emval_handles: "new HandleAllocator();",
   $emval_symbols: {}, // address -> string
 
-  $init_emval__deps: ['$count_emval_handles', '$get_first_emval'],
+  $init_emval__deps: ['$count_emval_handles', '$emval_handles'],
   $init_emval__postset: 'init_emval();',
   $init_emval: function() {
+#if ASSERTIONS
+    assert(emval_handles.allocated.length == 0);
+#endif
+    // reserve zero and special values, these never get de-allocated
+    emval_handles.allocated.push(
+      {},
```

Comment:
I wonder if zero should be reserved at HandleAllocator level - are there cases where we wouldn't want to catch 0/null-ish handle as special case?

Reply:
Agreed, yes.  Probably a separate change though.

Reply:
I ended up including that change, along with the one to avoid `delete` in the middle of an array.